core/src/consensus.rs

```diff
@@ -620,6 +720,215 @@ impl Tower {
+                last_voted_slot,
+                slot_history.oldest(),
+            ));
+        }
+
+        self.do_adjust_lockouts_after_replay(replayed_root_slot, slot_history)
+    }
+
+    fn do_adjust_lockouts_after_replay(
+        mut self,
+        replayed_root_slot: Slot,
+        slot_history: &SlotHistory,
+    ) -> Result<Self> {
+        // retained slots will be consisted only from divergent slots
+        let mut retain_flags_for_each_vote_in_reverse: Vec<_> =
+            Vec::with_capacity(self.lockouts.votes.len());
+
+        let mut still_in_future = true;
+        let mut past_outside_history = false;
+        let mut checked_slot = None;
+        let mut anchored_slot = None;
+
+        let mut root_checked = false;
+        let mut checked_slots = if let Some(root) = self.lockouts.root_slot {
+            root_checked = true;
+            vec![root]
+        } else {
+            vec![]
+        };
+        checked_slots.extend(self.voted_slots());
+
+        // iterate over votes in the newest => oldest order
+        // bail out early if bad condition is found
+        for voted_slot in checked_slots.iter().rev() {
+            checked_slot = Some(*voted_slot);
+            let check = slot_history.check(checked_slot.unwrap());
+
+            if anchored_slot.is_none() && check == Check::Found {
+                anchored_slot = checked_slot;
+            } else if anchored_slot.is_some() && check == Check::NotFound {
+                // this can't happen unless we're fed with bogus snapshot
+                return Err(TowerError::FatallyInconsistent("diverged ancestor?"));
+            }
+
+            if still_in_future && check != Check::Future {
+                still_in_future = false;
+            } else if !still_in_future && check == Check::Future {
+                // really odd cases: bad ordered votes?
+                return Err(TowerError::FatallyInconsistent("time warped?"));
+            }
+            if !past_outside_history && check == Check::TooOld {
+                past_outside_history = true;
+            } else if past_outside_history && check != Check::TooOld {
+                // really odd cases: bad ordered votes?
+                return Err(TowerError::FatallyInconsistent(
+                    "not too old once after got too old?",
+                ));
+            }
+
+            retain_flags_for_each_vote_in_reverse.push(anchored_slot.is_none());
+        }
+
+        assert_eq!(
+            checked_slots.len(),
+            retain_flags_for_each_vote_in_reverse.len()
+        );
+
+        if root_checked {
+            retain_flags_for_each_vote_in_reverse.pop();
+        }
+
+        // Check for errors if not anchored
+        // Being not anchored doesn't always equate to errors: All of votes in
+        // an unrooted (= !root_checked) warming-up vote account may legally not
+        // be anchored.
+        if anchored_slot.is_none() {
```

Comment:
actual error check for the not-anchored case: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR864-R868